File: core/packages/journal/src/journal-stream.ts
import { createReadStream, statSync } from "node:fs";
import { join } from "node:path";
import { createInterface } from "node:readline";
import { getJournalDirectory, listJournalFiles } from "./Journal.js";
import { parseLine } from "./parser.js";
import type { JournalEvent } from "./types.js";

export interface JournalStreamOptions {
  directory?: string;
  from?: "start" | "end";
  filter?: string | string[];
  pollInterval?: number;
}

export interface JournalStream {
  [Symbol.asyncIterator](): AsyncIterableIterator<JournalEvent>;
  stop(): void;
}

function sleep(ms: number, signal: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    const timer = setTimeout(resolve, ms);
    signal.addEventListener(
      "abort",
      () => {
        clearTimeout(timer);
        resolve();
      },
      { once: true },
    );
  });
}

async function* readFileEvents(
  file: string,
  startOffset: number,
  _endOffset: number,
  filterSet: Set<string> | null,
  signal: AbortSignal,
): AsyncIterableIterator<JournalEvent> {
  const stream = createReadStream(file, {
    start: startOffset,
    encoding: "utf-8",
  });
  const rl = createInterface({ input: stream, crlfDelay: Infinity });
  for await (const line of rl) {
    if (signal.aborted) return;
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const event = parseLine(trimmed);
      if (!filterSet || filterSet.has(event.event)) yield event;
    } catch {
      // skip malformed
    }
  }
}

export function createJournalStream(
  options: JournalStreamOptions = {},
): JournalStream {
  const dir = getJournalDirectory(options.directory);
  const filterSet = options.filter
    ? new Set(Array.isArray(options.filter) ? options.filter : [options.filter])
    : null;
  const pollInterval = options.pollInterval ?? 500;
  const abortController = new AbortController();
  const trackedSizes = new Map<string, number>();

  const stream: JournalStream = {
    async *[Symbol.asyncIterator]() {
      const currentFiles = listJournalFiles(dir);

      if (options.from !== "end") {
        for (const file of currentFiles) {
          if (abortController.signal.aborted) return;
          const size = statSync(file).size;
          trackedSizes.set(file, size);
          if (size <= 0) continue;
          yield* readFileEvents(
            file,
            0,
            size,
            filterSet,
            abortController.signal,
          );
        }
      } else {
        for (const file of currentFiles) {
          try {
            trackedSizes.set(file, statSync(file).size);
          } catch {
            // file vanished
          }
        }
      }

      while (!abortController.signal.aborted) {
        const files = listJournalFiles(dir);
        for (const file of files) {
          if (abortController.signal.aborted) return;

          const prevSize = trackedSizes.get(file) ?? 0;
          let currentSize: number;
          try {
            currentSize = statSync(file).size;
          } catch {
            continue;
          }

          if (currentSize > prevSize) {
            yield* readFileEvents(
              file,
              prevSize,
              currentSize,
              filterSet,
              abortController.signal,
            );
            trackedSizes.set(file, currentSize);
          } else if (currentSize < prevSize) {
            trackedSizes.set(file, currentSize);
          }
        }

        await sleep(pollInterval, abortController.signal);
      }
    },

    stop() {
      abortController.abort();
    },
  };

  return stream;
}

File: dotnet/EliteDangerousSdk/Journal/JournalStream.cs
using System.Runtime.CompilerServices;

namespace EliteDangerousSdk.Journal;

public record JournalStreamOptions
{
    public string? Directory { get; init; }
    public string From { get; init; } = "end";
    public string[]? Filter { get; init; }
    public int PollIntervalMs { get; init; } = 500;
    public bool WarnOnUnknown { get; init; }
}

public static class JournalStream
{
    public static async IAsyncEnumerable<Dictionary<string, object?>> CreateJournalStream(
        JournalStreamOptions? options = null,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var dir = options?.Directory ?? JournalReader.GetDefaultJournalDir();
        var filterSet = options?.Filter is { Length: > 0 } ? new HashSet<string>(options.Filter) : null;
        var pollInterval = options?.PollIntervalMs ?? 500;
        var warnOnUnknown = options?.WarnOnUnknown ?? false;
        var trackedSizes = new Dictionary<string, long>();

        var files = JournalReader.ListJournalFiles(dir);

        // Phase 1: Catch-up
        if (options?.From != "end")
        {
            foreach (var file in files)
            {
                if (ct.IsCancellationRequested) yield break;
                var size = new FileInfo(file).Length;
                trackedSizes[file] = size;
                if (size <= 0) continue;

                using var stream = File.OpenRead(file);
                using var reader = new StreamReader(stream);
                string? line;
                while ((line = reader.ReadLine()) != null)
                {
                    if (ct.IsCancellationRequested) yield break;
                    line = line.Trim();
                    if (string.IsNullOrEmpty(line)) continue;
                    Dictionary<string, object?>? ev = null;
                    try
                    {
                        ev = Parser.ParseLine(line);
                    }
                    catch { }
                    if (ev != null && (filterSet == null || filterSet.Contains(ev.GetValueOrDefault("event") as string ?? "")))
                    {
                        JournalReader.WarnIfUnknown(ev, warnOnUnknown);
                        yield return ev;
                    }
                }
            }
        }
        else
        {
            foreach (var file in files)
            {
                try { trackedSizes[file] = new FileInfo(file).Length; }
                catch { }
            }
        }

        // Phase 2: Live polling
        while (!ct.IsCancellationRequested)
        {
            var currentFiles = JournalReader.ListJournalFiles(dir);
            foreach (var file in currentFiles)
            {
                if (ct.IsCancellationRequested) yield break;

                var prevSize = trackedSizes.GetValueOrDefault(file, 0L);
                long currentSize;
                try { currentSize = new FileInfo(file).Length; }
                catch { continue; }

                if (currentSize > prevSize)
                {
                    using var fs = File.OpenRead(file);
                    fs.Seek(prevSize, SeekOrigin.Begin);
                    using var sr = new StreamReader(fs);
                    string? line;
                    while ((line = sr.ReadLine()) != null)
                    {
                        if (ct.IsCancellationRequested) yield break;
                        line = line.Trim();
                        if (string.IsNullOrEmpty(line)) continue;
                        Dictionary<string, object?>? ev = null;
                        try
                        {
                            ev = Parser.ParseLine(line);
                        }
                        catch { }
                        if (ev != null && (filterSet == null || filterSet.Contains(ev.GetValueOrDefault("event") as string ?? "")))
                        {
                            JournalReader.WarnIfUnknown(ev, warnOnUnknown);
                            yield return ev;
                        }
                    }
                    trackedSizes[file] = fs.Position;
                }
                else if (currentSize < prevSize)
                {
                    trackedSizes[file] = currentSize;
                }
            }

            try { await Task.Delay(pollInterval, ct); }
            catch (OperationCanceledException) { break; }
        }
    }
}
